snorkel/slicing/utils.py

```diff
@@ -0,0 +1,202 @@
+    away in the model since they have the same name. We leave resolution of this issue
+    for a future release.
+    """
+    if "base" not in slice_names:
+        slice_names = slice_names + ["base"]
+
+    tasks = []
+
+    # NOTE: We assume here that the last operation uses the head module
+    # Identify base task head module
+    head_module_op = base_task.task_flow[-1]
+    head_module = base_task.module_pool[head_module_op.module_name]
+
+    if isinstance(head_module, nn.DataParallel):
+        head_module = head_module.module
+
+    neck_size = head_module.in_features
+    base_task_cardinality = head_module.out_features
+
+    # Remove the slice-unaware head module from module pool and task flow
+    del base_task.module_pool[head_module_op.module_name]
+    base_task.task_flow = base_task.task_flow[:-1]
+
+    # Create slice indicator tasks
+    for slice_name in slice_names:
+
+        ind_task_name = f"{base_task.name}_slice:{slice_name}_ind"
+
+        ind_head_module_name = f"{ind_task_name}_ind_head"
+        ind_head_module = nn.Linear(neck_size, 2)
+
+        # Create module_pool
+        ind_module_pool = base_task.module_pool
+        ind_module_pool[ind_head_module_name] = ind_head_module
+
+        # Create task_flow
+        ind_head_op = Operation(
+            module_name=ind_head_module_name, inputs=head_module_op.inputs
+        )
+        ind_task_flow = base_task.task_flow + [ind_head_op]
+
+        # Create ind task
+        ind_task = Task(
+            name=ind_task_name,
+            module_pool=ind_module_pool,
+            task_flow=ind_task_flow,
+            loss_func=partial(ce_loss, ind_head_op.name),
+            output_func=partial(softmax, ind_head_op.name),
+            scorer=Scorer(metrics=["accuracy"]),
+        )
+        tasks.append(ind_task)
+
+    # Create slice predictor tasks
+    shared_pred_head_module = nn.Linear(neck_size, base_task_cardinality)
+    for slice_name in slice_names:
+
+        pred_task_name = f"{base_task.name}_slice:{slice_name}_pred"
+
+        pred_head_module_name = f"{pred_task_name}_pred_head"
+        pred_transform_module_name = f"{pred_task_name}_pred_transform"
+        pred_transform_module = nn.Linear(neck_size, neck_size)
+
+        # Create module_pool
+        # NOTE: See note in doc string about module_pool polution
+        pred_module_pool = base_task.module_pool
+        pred_module_pool[pred_transform_module_name] = pred_transform_module
+        pred_module_pool[pred_head_module_name] = shared_pred_head_module
+
+        # Create task_flow
+        pred_transform_op = Operation(
+            module_name=pred_transform_module_name, inputs=head_module_op.inputs
+        )
+        pred_head_op = Operation(
+            module_name=pred_head_module_name, inputs=[(pred_transform_op.name, 0)]
+        )
+        pred_task_flow = base_task.task_flow + [pred_transform_op, pred_head_op]
```

Comment:
same as above—temporarily confusing that we're still calling the decapitated base_task flow by the same name